Revamp Swarm Control Package

hungjn opened the pull request:
- Change the swarm control package to run a routine which controls a swarm of rovers to level an area on the Moon map.

hungjn committed:
Fsbp 52 MR (#2)

* Move rover functionality to the rover class.
Use rover class within the main while loop.

* Add test message.

* fixup

* Changed pos to pose

* Added self to dig_sites line 173

Co-authored-by: CoyTorreblanca <coydiego@knights.ucf.edu>

Coy-Torreblanca committed:
Fsbp 40 (#3)


* Add dig command to swarm control constants.

* Add dump command to waypoint client.

* Add auto dig function with waypoint support.

* Add dump command to waypoint server.

* Add dump method to rover class.

Co-authored-by: CoyTorreblanca <coydiego@knights.ucf.edu>
Co-authored-by: Hung <hungjn748@gmail.com>

Coy-Torreblanca committed:
Fsbp 55 (#4)

* Add leveling algorithm

* Added imports for level algo

Co-authored-by: CoyTorreblanca <coydiego@knights.ucf.edu>
Co-authored-by: Hung <hungjn748@gmail.com>

hungjn committed:
Added image processing code (#5)

Coy-Torreblanca committed:
Dig and dump sites are only considered if their elevation is above or below a given threshold. (#6)

Co-authored-by: CoyTorreblanca <coydiego@knights.ucf.edu>

hungjn committed:
Fsbp 59 MR (#7)

* Added pixel_normalize and pixel_expand methods to swarm utils

* Imported pixel normalize to swarm control

hungjn committed:
Convert image coordinates to gazebo simulation coordinates in leveling algorithm output. (#8)

Co-authored-by: CoyTorreblanca <coydiego@knights.ucf.edu>

hungjn committed:
Fsbp 46 (#13)

* Add rover class to control rovers.

    Rover class has instance variables which describe rover properties, including a thread object.
    Rover class has instance methods which perform basic rover functionality, e.g. get_rover_status.
    Add Rover class functionality that pertains to build pad.
    Add Rover object ability to ask Swarm Controller for instructions.
    Add to Rover class the level function which commands rover to dig and dump between two locations.
        Rover object will stop leveling once all instructed actions have been acomplished.
    Rover object threads will spin until it recieves an instruction from the SwarmController.
    Add Rover class go to function which handles getting a rover from its current destination to a target destination.

* Add Swarm Controller functionality to manage rovers.

    Swarm controller assigns rovers dig dump pairs.
    Add data structure to track actions required for each dig and dump pair.
    Add data structures to track which dig and dump locations are assigned to rovers.
    Rovers update global data structure. Rovers use locks to prevent thread race conditions.
    Swarm controller stops when map is level or when all locations left to level are blacklisted.

* Rovers calculate battery requirements.

    Rovers go and charge if they do not have enough battery to complete one assigned action.
    Rovers complete the number of actions that they have enough battery for.
    Rovers will go and charge if they were not able to complete all of the actions required by assignment.
    Rovers will update central data structures when they preempt or finish an assignment.
    Add battery checking after each sub-action in rover level function.
    Add padding to battery checking and battery calculations.
    Front end believes that the rover has 10% less battery to provide a buffer.

* Allow waypoint client to send messages to rover objects.

    Add a ROS subscriber to Rover node which gets messages from waypoint client when an action has been completed.
    The type of action completed is described by the enum Action codes.
    Rovers will wait to continue their routine until the correct integer is published by the waypoint client.


* Waypoint client will preempt pathing more often.

    Lower max veer distance before a new path is calculated in an attempt to
    stop rovers from moving too far away from desired location.


* Handle when rover routine encounters problems.

    Add rover ability to abort assignment when pathing and communicate problem to swarm controller. Swarm controller will handle rover problems.
    Rovers will diagnose problem before raising an exception.
    Create a custom exception for when rover aborts action sequence.
    Swarm controller will understand problem through RoverState enum type
    instance variable in Rover object.
    Swarm controller will blacklist problem locations so that other rovers will avoid them.

* Skip possible instructions which contain a location which have caused a rover problem.

* Crop original height map to represent the desired leveling area.

* Add landing pad auto-functions.

    Add new auto-dig, auto-dump, auto-drive-location which is specialized for the build landing pad swarm.
    Add variable to world state (carrying_dirt) which is set to true after auto-dig-land-pad is called and set to false when auto-dump-land-pad is called.
    Add extra cost to auto-drive-location_land_pad when rover is carrying dirt.
    Auto-dig-land-pad and auto-dump-land-pad turn to the same absolute angle before every execution.
    Add battery cost to auto-dump-land-pad.
    Moved additional functions to the bottom of the auto_functions file.

* Reformat to Black formatting.

* Implement a cleaner and faster leveling algorithm.

    Has the same logic as the original leveling algorithm.

* Lower obstacle buffer in an attempt to make obstacle detection less strict.
* Change dig and dump time to ensure that the rover begins and ends at the same location.

Co-authored-by: hungjn <67024640+hungjn@users.noreply.github.com>
Co-authored-by: CoyTorreblanca <coydiego@knights.ucf.edu>
Co-authored-by: Coy-Torreblanca <78828209+Coy-Torreblanca@users.noreply.github.com>
Co-authored-by: Stanley <staanleyminervini@gmail.com>

CoyTorreblanca committed:
squash! Fsbp 46 (#13)

CoyTorreblanca committed:
squash! Fsbp 46 (#13)

hungjn committed:
Added logging of blacklist sites at end of simulation

hungjn committed:
Added logging of blacklist sites at end of simulation

CoyTorreblanca committed:
Change formatting to satisfy linter.

hungjn committed:
Added immobilized rover data structure and logging

hungjn committed:
Added immobilized rover data structure and logging

hungjn committed:
Added conditional checking for end of simulation logging

hungjn committed:
Added conditional checking for end of simulation logging

CoyTorreblanca committed:
Change formatting to satisfy linter.

CoyTorreblanca committed:
squash! Change formatting to satisfy linter.

Merge branch 'mainline' of github.com:hungjn/EZ-RASSOR into mainline

# Conflicts:
#	packages/autonomy/ezrassor_swarm_control/source/ezrassor_swarm_control/swarm_control.py

hungjn committed:
Fix PEP8 linter formatting

hungjn committed:
Fix PEP8 linter formatting

CoyTorreblanca committed:
Rover will move while dumping in auto landing pad.

CoyTorreblanca committed:
Rover will move while dumping in auto landing pad.

CoyTorreblanca committed:
Rover will move while dumping in auto landing pad.

Rover will stop before changing directions while digging/dumping to stop
drastic rover motion.

CoyTorreblanca committed:
Lower battery when rover sends message to SwarmC.

CoyTorreblanca committed:
Fix for edge case for when all rovers are stuck.

Add a log for when the entire area is leveled.

CoyTorreblanca committed:
Fix for edge case for when all rovers are stuck.

Add a log for when the entire area is leveled.

CoyTorreblanca committed:
Send rover home after area is level.

CSharpRon commented on packages/autonomy/ezrassor_autonomous_control/source/ezrassor_autonomous_control/utility_functions.py:
```diff
@@ -165,7 +165,7 @@ def move(dist, world_state, ros_util, direction="forward"):
             ros_util.obstacle_threshold,
         ):
-            rospy.loginfo("Obstacle too close! Stopping!")
+            # rospy.loginfo("Obstacle too close! Stopping!")
```
Any reason why this is commented out? If it's redundant then you should just nix it.

Coy-Torreblanca replied:
We commented it out for testing purposes. We will revert the change.

CSharpRon commented on packages/autonomy/ezrassor_autonomous_control/source/ezrassor_autonomous_control/utility_functions.py:
```diff
@@ -287,10 +287,10 @@ def get_turn_angle(world_state, ros_util):
-                "There is nowhere to go in the current wedge. "
-                + "Turning to an adjacent wedge."
-            )
+            # rospy.loginfo(
```
Remove if redundant.

Coy-Torreblanca replied:
We commented it out for testing purposes. We will revert the change.

CSharpRon commented on packages/autonomy/ezrassor_autonomous_control/source/ezrassor_autonomous_control/utility_functions.py:
```diff
@@ -321,7 +321,7 @@ def get_turn_angle(world_state, ros_util):
                 rospy.sleep(0.1)
 
-                rospy.loginfo("Currently at wedge W{}".format(wedgeDist - 1))
+                # rospy.loginfo("Currently at wedge W{}".format(wedgeDist - 1))
```
Remove if redundant

Coy-Torreblanca replied:
We commented it out for testing purposes. We will revert the change.

CSharpRon commented on packages/autonomy/ezrassor_swarm_control/source/ezrassor_swarm_control/swarm_utils.py:
```diff
@@ -68,3 +71,269 @@ def preempt_rover_path(rover_num):
+    standard_dig_amount = 1
+
+    # loop through array and standardize the array
+    for row in range(0, img_width):
```
This should make use of numpy vectorization which is much faster than a for-loop approach. To accomplish the below you could accomplish it with:
```py
height_matrix = (height_matrix - level_elevation) / standard_dig_amount
```
given that height_matrix is a numpy array, and level_elevation + standard_dig_amount are both ints or floats.

CSharpRon commented on packages/autonomy/ezrassor_swarm_control/source/ezrassor_swarm_control/swarm_utils.py:
```diff
@@ -68,3 +71,269 @@ def preempt_rover_path(rover_num):
+    scaled_array_height = img_height * scale_height
+    scaled_array = np.zeros((scaled_array_height, scaled_array_width))
+
+    for row in range(0, img_width):
```
I'm not quite sure what's going on here but if the output is a scaled image, cv2 has functions that can help with that like cv2.resize.

Coy-Torreblanca replied:
I've taken a look at cv2 resize, and I cannot tell if it can do what we need.
We will keep taking a look at it.

CSharpRon commented:
Great work so far with the PR. Make sure to install and run black on your cloned repository so that you can pass the lint test next time. Also, make sure that "squash and merge" is selected when the PR is final so that only one commit will get added to Mainline.

CoyTorreblanca committed:
Rovers will stop when digging/dumping...

Before changing changing directions in auto_dump_land_pad and
auto_dig_land_pad, make sure the rover comes to a complete stop to avoid
the rover tipping to drastically.

CoyTorreblanca committed:
Change autonomous variables for stable driving.

CoyTorreblanca committed:
Change autonomous variables for stable driving.

CoyTorreblanca committed:
Change dump battery cost estimate.

To reflect that the rover moves while dumping.

CoyTorreblanca committed:
Change dump battery cost estimate.

To reflect that the rover moves while dumping.

CoyTorreblanca committed:
Merge branch 'dump_move' into mainline

StanleyMine committed:
utilize numpy vectorization

CoyTorreblanca committed:
Add feature to avoid rover collisions.

Do not assign dig/dump locations which are near other assigned locations
to prevent rovers from colliding during digging/dumping.

CoyTorreblanca committed:
Merge branch 'battery_change_msgs' into mainline

CoyTorreblanca committed:
fixup! Add feature to avoid rover collisions.

CoyTorreblanca committed:
Merge branch 'rover_collisions' into mainline

CoyTorreblanca committed:
Fix inaccurate battery calculation variable.

CoyTorreblanca committed:
Communicate to supervisor if rover is immobile.

CoyTorreblanca committed:
Lower rover arms when dumping to prevent wheelies.

CoyTorreblanca committed:
fixup! Fix inaccurate battery calculation variable.

CoyTorreblanca committed:
fixup! fixup! Fix inaccurate battery calculation variable.

CoyTorreblanca committed:
fixup! fixup! fixup! Fix inaccurate battery calculation variable.

CoyTorreblanca committed:
Rovers will charge where they originaly spawned.

CoyTorreblanca committed:
fixup! Rovers will charge where they originaly spawned.

CoyTorreblanca committed:
Ignore flake E203 on swarm_control.

Flake throws E203 on line formatted by black.

CoyTorreblanca committed:
Remove battery loss due to messages.

We cannot accurately estimate how much battery is required by sending
messages in swarm_control; therefore, the rover runs out of battery in
the simulation.

CoyTorreblanca committed:
Uncomment necessary logging.

StanleyMine committed:
rewrite image processing function

StanleyMine committed:
Merge branch 'mainline' of github.com:hungjn/EZ-RASSOR into mainline

hungjn committed:
Added image processed dem

CoyTorreblanca committed:
Change to black format.

hungjn committed:
Updated readme team information

hungjn committed:
Added newline after team name

CoyTorreblanca committed:
Rover charges before custom routine.

CoyTorreblanca committed:
Merge branch 'mainline' of github.com:hungjn/EZ-RASSOR into mainline

StanleyMine committed:
Fix sleep time for dumping

CoyTorreblanca committed:
Add documentation.

CoyTorreblanca committed:
Add Unity sim.

CoyTorreblanca committed:
Revert "Add Unity sim" because build failed.

This reverts commit fc96f1b092249b26575a435d95ecc2bcc2b97318.

Coy-Torreblanca commented:
If preferred, I can revert this branch to the commit where we resolved Ron's comments and add the other new changes to a new pull request.

CSharpRon commented on packages/autonomy/ezrassor_swarm_control/source/ezrassor_swarm_control/swarm_utils.py:
```diff
@@ -68,3 +71,225 @@ def preempt_rover_path(rover_num):
+        dig_dump_pairs[dig_location].append((dump_location, elevation_change))
+
+    # Convert image coordinates to simulation coordinates in a Point object.
+    def convert_to_simulation_Point(matrix, x, y):
```
Will this package fail if the simulation is not running? If so, then logic like this should be abstracted to a sim-specific package so that this swarm control code can be arbitrary enough to run on either hardware or sim.

Coy-Torreblanca replied:
We convert the rover destination instructions into simulation format because the A* pathing algorithm and the autonomous package require rover destinations to be in simulation format. 

What we could do is make the A* planner take regular format destinations and then convert them into simulation coordinates before sending them over to the autonomous package. Should I make an issue? I can make the change in the future.

CSharpRon reviewed (approved):
Fantastic updates since the last review. I added one comment for review but am still approving the PR. Also, there are ALOT of commits in this PR. It's not a problem in itself because that's how development works ofc, but I would like to have this entire PR squashed as one commit when it is merged back into the main branch.

Before checking the box and squashing the PR as a single commit, I recommend looking up how to add co-authors to a squashed commit in GitHub. That way everyone who worked on these changes will get credit (as opposed to just the author of the last commit which I think is the default).
